Feature/add concurrency task

yayauheny opened the pull request.

yayauheny committed:
feature: add classes

All changes:
- Add Client, Server, Request, Response, ConcurrencyRunner classes
- Add main logic to the classes
- ConcurrencyRunner class initialization

yayauheny committed:
feature: add main logic

All changes:
- Refactor Client and Server classes for stability and cleaner code
  view
- Add logic in ConcurrencyRunner for correct work

yayauheny committed:
refactor: refactor main method

yayauheny committed:
refactor: move logic

All changes:
- Move logic from ConcurrencyRunner to Client and Server
- Create additional ReentrantLocks
- Use streams

yayauheny committed:
refactor: optimize code

yayauheny committed:
feature: fix bugs and add tests

Lexuspavluk commented on src/main/java/com/console/ticket/concurrency/Request.java:
```diff
@@ -0,0 +1,14 @@
+
+    private int value;
+
+    public int getValue() {
```
Переходи уже на lombok или record-class

Lexuspavluk commented on src/main/java/com/console/ticket/concurrency/Client.java:
```diff
@@ -0,0 +1,134 @@
+        this.sendData = sendData;
+    }
+
+    public int getAccumulator() {
```
Если слушал лекцию - Atomic - не лочатся. Они синхронизируются по другому принципу (CAS-compareAndSet)

Lexuspavluk commented on src/main/java/com/console/ticket/concurrency/Client.java:
```diff
@@ -0,0 +1,134 @@
+
+    public void start(Server server, int threadsQuantity, int queueSize) {
+        clientExecutor = Executors.newFixedThreadPool(threadsQuantity);
+        requestsFromClientQueue = new LinkedBlockingQueue<>(queueSize);
```
Такие инициализации в конструкторе делают, или в блоке инициализации. К логике start отношения не имеют.

Lexuspavluk commented on src/main/java/com/console/ticket/concurrency/ConcurrencyRunner.java:
```diff
@@ -0,0 +1,33 @@
+        System.out.println(client.getAccumulator());
+    }
+
+    private static CopyOnWriteArrayList<Integer> getIntegerQueue(int size){
```
что-то с неймингом не то... Queue или List ?

yayauheny replied:
Перебирал коллекции, все не то, а перед отправкой забыл поменять название)

Lexuspavluk commented on src/main/java/com/console/ticket/concurrency/ConcurrencyRunner.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    private static CopyOnWriteArrayList<Integer> getIntegerQueue(int size){
+        CopyOnWriteArrayList<Integer> integerQueue = new CopyOnWriteArrayList<>();
```
Прикольно... но не забывай про KISS

Lexuspavluk commented on src/main/java/com/console/ticket/concurrency/Server.java:
```diff
@@ -0,0 +1,51 @@
+        this.receivedData = receivedData;
+    }
+
+    public void addElement(Request request) {
```
методы сделать приватными и наружу выставить метод принимающий Request и возвращающий Response

Lexuspavluk commented on src/main/java/com/console/ticket/concurrency/Server.java:
```diff
@@ -0,0 +1,51 @@
+        int currentSize;
+
+        try {
+            receivedDataLock.lock();
```
...и между добавлением значения в лист и получением его размера прилетит еще один Request и успеет внести свое значение. Лок на эти операции должен быть один.

yayauheny replied:
Так у меня же 1 лок на все операции с листом)

yayauheny replied:
Так вот в чем была проблема.. Я вынес их в 1 метод и навесил лок - и теперь все работает как часы, спасибо)

Lexuspavluk commented on src/main/java/com/console/ticket/concurrency/Client.java:
```diff
@@ -0,0 +1,134 @@
+    private void putAllResponsesFromServer(Server server, Queue<Future<Request>> requestsQueue, Queue<Future<Response>> responsesQueue) {
+        Stream.generate(requestsQueue::poll)
+                .takeWhile(Objects::nonNull)
+                .forEach(requestFuture -> {
```
Неудобочитаемо. Выноси такие объемные блоки кода из стримов в отдельный метод.

Lexuspavluk commented on src/main/java/com/console/ticket/concurrency/Client.java:
```diff
@@ -0,0 +1,134 @@
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+public class Client implements Callable<Request> {
```
Сервер сам по себе однопоточный. Нет смысла его имплементировать от Callable

yayauheny committed:
refactor: modify logic and refactor classes

All changes:
- Make Request and Response classes as records
- Move initialization from start() method to constructor
- Rename methods
- Delete redundant lock for AtomicInteger
- Replace redundant method getIntegerList() with initialisation with Stream
- Simplify and rename methods
- Refactor tests
- Move 2 operations into 1 public method handleRequestAndReturnDataSize() and lock it

Lexuspavluk reviewed (commented):
Норм.
